### PR TITLE
show original price in listing + product sheet

### LINKED ISSUE #8258: Original price not displayed in product listing
Create a product variant with price and original price.
On product listing only the price is being displayed, original price should also be displayed crossed out.

I'm ready to contribute if you see it's importance.

### PR DESCRIPTION
**Pull request by Snake-Tn**
| Q               | A
| --------------- | ---
| Bug fix?        | yes |
| New feature?    | no |
| BC breaks?      | yes |
| Related tickets | fixes #8258 |
| License         | MIT |

### CODE REVIEW

File: features/product/viewing_products/viewing_products_original_price.feature [hunk 1]
```diff
@@ -0,0 +1,14 @@
+@viewing_products
+Feature: Viewing a product original price
+    In order to see products original price
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_products_original_price.feature]: `In order to be informed about the amount of discount`?

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 2]
```diff
@@ -199,6 +199,16 @@ public function storeHasAProductPricedAt($productName, $price = 100, ChannelInte
     }
 
     /**
+     * @Given /^the store(?:| also) has a product "([^"]+)" originally priced at ("[^"]+")$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: Do we need so generic step?

File: features/product/viewing_products/viewing_products_original_price.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    As a Visitor
+    I want to be able to view a single product original price
+
+    Background:
+        Given the store operates on a single channel in "United States"
+
+    @ui
+    Scenario: Viewing a detailed page with product's original price
+        Given the store has a product "T-shirt banana" originally priced at "$59.00"
```
Comment (lchrusciel) [inline features/product/viewing_products/viewing_products_original_price.feature]:
```gherkin
Given the store has a product "T-shirt banana" priced at "$49.00"
And this product has been originally priced at "$59.00"
```
?

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 4]
```diff
@@ -199,6 +199,16 @@ public function storeHasAProductPricedAt($productName, $price = 100, ChannelInte
...
+     */
+    public function storeHasAProductOriginallyPricedAt($productName, $originalPrice = 100, ChannelInterface $channel = null)
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: The channel parameter is not needed. 

Also, if you will use the proposal from one comment above, we can use a shared storage transformer.
```php
     /**
       * @Given /^(this product) has been originally priced at ("[^"]+")$/
       */
      public function thisProductIsAlsoPricedAtInChannel(ProductInterface $product, $price)
     {
         /** @var ProductVariantInterface $productVariant */
         $productVariant = $this->defaultVariantResolver->getVariant($product);

         // For all channel pricings for example
 
         $this->objectManager->flush();
     }
```

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php [hunk 5]
```diff
@@ -54,6 +63,19 @@ public function getPrice(ProductVariantInterface $productVariant, array $context
...
      * {@inheritdoc}
      */
+    public function getOriginalPrice(ProductVariantInterface $productVariant, array $context): int
+    {
+        Assert::keyExists($context, 'channel');
+
+        return $this
+            ->productVariantOriginalPriceCalculator
+            ->calculate($productVariant, $context)
+            ;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: `;` should be on level of `return`

**Comment by Snake-Tn** [comment]
@lchrusciel , Thanks for your help.
I fixed issues related to behat features as you mentioned. 
Can you re-check please ?

**Comment by lchrusciel** [comment]
@Snake-Tn I will try to do it in nearest days :)

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php [hunk 6]
```diff
@@ -29,11 +29,20 @@ class PriceHelper extends Helper
...
+
+    /**
      * @param ProductVariantPriceCalculatorInterface $productVariantPriceCalculator
+     * @param ProductVariantPriceCalculatorInterface $productVariantOriginalPriceCalculator
      */
-    public function __construct(ProductVariantPriceCalculatorInterface $productVariantPriceCalculator)
-    {
+    public function __construct(
+        ProductVariantPriceCalculatorInterface $productVariantPriceCalculator,
+        ProductVariantPriceCalculatorInterface $productVariantOriginalPriceCalculator
```
Comment (lchrusciel) [inline src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: I would create a new interface to split the logic

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig [hunk 7]
```diff
@@ -1,5 +1,8 @@
 {% import "@SyliusShop/Common/Macro/money.html.twig" as money %}
 
+<span class="ui small header" style="text-decoration: line-through" id="product-original-price">
```
Comment (lchrusciel) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig]: Can we determine if we even want to add this html block?

File: src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php [hunk 8]
```diff
@@ -25,7 +25,7 @@
     /**
      * {@inheritdoc}
      */
-    public function calculate(ProductVariantInterface $productVariant, array $context): int
+    public function calculate(ProductVariantInterface $productVariant, array $context): ?int
```
Comment (lchrusciel) [inline src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php]: Can we revert this change?

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php (hunk 6 above)
Comment (Snake-Tn) [reply src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: Yes it would be better since original price is nullble and price is not. So they should not share the same calculator interface .

File: src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php (hunk 8 above)
Comment (Snake-Tn) [reply src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php]: Yes, after creating a new interface for original price .

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig (hunk 7 above)
Comment (Snake-Tn) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig]: You mean in case original price is null or higher than the price, we should not print this HTML?
Currently this login is hold be a twig macro.

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php [hunk 9]
```diff
@@ -318,6 +326,7 @@ protected function getDefinedElements()
             'option_select' => '#sylius_add_to_cart_cartItem_variant_%option-name%',
             'out_of_stock' => '#sylius-product-out-of-stock',
             'product_price' => '#product-price',
+            'product_original_price' => '#product-original-price',
```
Comment (GSadee) [inline src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: Wrong order in the array

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php [hunk 10]
```diff
@@ -54,6 +63,18 @@ public function getPrice(ProductVariantInterface $productVariant, array $context
     /**
      * {@inheritdoc}
      */
+    public function getOriginalPrice(ProductVariantInterface $productVariant, array $context): ?int
+    {
+        Assert::keyExists($context, 'channel');
+
+        return $this
+            ->productVariantOriginalPriceCalculator
+            ->calculate($productVariant, $context);
```
Comment (GSadee) [inline src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: Move `;` to the next line

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php (hunk 6 above)
Comment (gabiudrescu) [reply src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: why not set the original price equal to selling price and respect the same interface?

Comment (Snake-Tn) [reply src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: I'm agree that null values are no more than a door for bugs to come from. And want to avoid them as possible. In best practice a method should not has multi return type. 
It's possible to update the way how a product get persisted to assure original price has an initial value(equal price). Then there is the problem of data migration to deal with also.(in order to update current products)
This is big decisions should be validated with sylius maintainers I think .

Comment (Snake-Tn) [reply src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: I'm agree that null values are no more than a door for bugs to come from. And want to avoid them as possible. In best practice a method should not has multi return type. 
It's possible to update the way how a product get persisted to assure original price has an initial value(equal price). Then there is the problem of data migration to deal with also.(in order to update current products)
This is big decisions should be validated with sylius maintainers I think .

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php (hunk 9 above)
Comment (Snake-Tn) [reply src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: Why order is wrong , what is the correct order ?

**Comment by Snake-Tn** [comment]
@lchrusciel , regarding price/original price calculator. 
We have two options : 

1. Create a separate interface for original price.
2. Original price is per default equal to the price [@gabiudrescu idea] (then original price will never be null, and it can share same interface with price)

What option do you prefer to implement ?

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php (hunk 9 above)
Comment (GSadee) [reply src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: `product_original_price` key should be before `product_price` key :)

File: src/Sylius/Behat/Context/Setup/ProductContext.php [hunk 11]
```diff
@@ -197,6 +197,16 @@ public function storeHasAProductPricedAt($productName, $price = 100, ChannelInte
     }
 
     /**
+     * @Given /^(this product) has been originally priced at ("[^"]+")$/
+     */
+    public function thisProductIsAlsoOriginallyPricedAtInChannel(ProductInterface $product, $originalPrice = 100, ChannelInterface $channel = null)
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/ProductContext.php]: `ChannelInterface $channel = null` is not needed.

Comment (Zales0123) [reply src/Sylius/Behat/Context/Setup/ProductContext.php]: Also, although it's they're not used widely in **Behat** yet, we should start using strict scalar types, as we support only PHP 7.1 -> so `int $originalPrice` (I don't think the default value is needed) and `: void` at the end.

File: src/Sylius/Behat/Context/Ui/Shop/ProductContext.php [hunk 12]
```diff
@@ -292,6 +292,14 @@ public function iShouldSeeTheProductPrice($price)
     }
 
     /**
+     * @Then I should see the product original price :originalPrice
+     */
+    public function iShouldSeeTheProductOriginalPrice($originalPrice)
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/ProductContext.php]: Same here, and everywhere else ;)

File: src/Sylius/Behat/Page/Shop/Product/ShowPage.php (hunk 9 above)
Comment (Zales0123) [reply src/Sylius/Behat/Page/Shop/Product/ShowPage.php]: Alphabetical order is life 🔤

File: src/Sylius/Bundle/ShopBundle/Resources/views/Common/Macro/money.html.twig [hunk 13]
```diff
@@ -13,3 +13,12 @@
 
     {{- convertAndFormat(variant|sylius_calculate_price({'channel': sylius.channel})) }}
 {%- endmacro -%}
+
+{%- macro calculateOriginalPrice(variant) -%}
+    {% from _self import convertAndFormat %}
+    {% set originalPrice = variant|sylius_calculate_original_price({'channel': sylius.channel})  %}
+    {% set price = variant|sylius_calculate_price({'channel': sylius.channel})  %}
+    {% if( originalPrice > price) %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Common/Macro/money.html.twig]: Missing space after `if` and additional space before `originalPrice`

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig (hunk 7 above)
Comment (Zales0123) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig]: I think this condition should be put here - we don't need a redundant html with no text in it :)

File: src/Sylius/Component/Core/spec/Calculator/ProductVariantOriginalPriceCalculatorSpec.php [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+namespace spec\Sylius\Component\Core\Calculator;
+
+use Sylius\Component\Core\Calculator\ProductVariantOriginalPriceCalculator;
+
+/**
+ * @author Ahmed Kooli <kooliahmd@gmail.com>
+ */
+class ProductVariantOriginalPriceCalculatorSpec
+{
+    function it_is_initializable()
```
Comment (Zales0123) [inline src/Sylius/Component/Core/spec/Calculator/ProductVariantOriginalPriceCalculatorSpec.php]: We've decided not to use `it_is_initializable()` functions some time ago

File: src/Sylius/Component/Core/spec/Calculator/ProductVariantOriginalPriceCalculatorSpec.php [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+ */
+
+namespace spec\Sylius\Component\Core\Calculator;
+
+use Sylius\Component\Core\Calculator\ProductVariantOriginalPriceCalculator;
+
+/**
+ * @author Ahmed Kooli <kooliahmd@gmail.com>
+ */
+class ProductVariantOriginalPriceCalculatorSpec
```
Comment (Zales0123) [inline src/Sylius/Component/Core/spec/Calculator/ProductVariantOriginalPriceCalculatorSpec.php]: Missing `final`

File: src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php (hunk 6 above)
Comment (Zales0123) [reply src/Sylius/Bundle/CoreBundle/Templating/Helper/PriceHelper.php]: Even though we decide to implement this feature, still `$originalPrice` and `$price` are slightly different concepts and I think it should be useful to have different interfaces for their calculators.

**Comment by gabiudrescu** [comment]
kind reminder

File: src/Sylius/Bundle/ShopBundle/Resources/views/Common/Macro/money.html.twig [hunk 16]
```diff
@@ -13,3 +13,12 @@
...
+        {{- convertAndFormat(variant|sylius_calculate_original_price({'channel': sylius.channel})) }}
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Common/Macro/money.html.twig]: `{{- convertAndFormat(originalPrice) -}}`, no need to calculate it twice

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig (hunk 7 above)
Comment (pamil) [reply src/Sylius/Bundle/ShopBundle/Resources/views/Product/Show/_price.html.twig]:
```
{% set originalPrice = money.calculateOriginalPrice(product|sylius_resolve_variant) %}
{% if originalPrice is not null %}
    <span>{{ originalPrice }} </span>
{% endif %}
```

Should do the work

File: src/Sylius/Bundle/ShopBundle/Resources/views/Product/_box.html.twig [hunk 17]
```diff
@@ -16,6 +16,7 @@
     <div class="content">
         <a href="{{ path('sylius_shop_product_show', {'slug': product.slug, '_locale': product.translation.locale}) }}" class="header sylius-product-name">{{ product.name }}</a>
         {% if not product.variants.empty() %}
+            <div class="sylius-product-original-price" style="text-decoration: line-through">{{ money.calculateOriginalPrice(product|sylius_resolve_variant) }}</div>
```
Comment (pamil) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Product/_box.html.twig]: Same here as above (with the excess HTML being generated)

File: src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php (hunk 8 above)
Comment (pamil) [reply src/Sylius/Component/Core/Calculator/ProductVariantPriceCalculator.php]: Yeah, this needs to be reverted in order to keep BC.

File: src/Sylius/Component/Core/Calculator/ProductVariantOriginalPriceCalculator.php [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+
+        if (null === $channelPricing) {
+            throw new MissingChannelConfigurationException(sprintf(
+                'Channel %s has no price defined for product variant %s',
+                $context['channel']->getName(),
+                $productVariant->getName()
+            ));
+        }
+
+        return $channelPricing->getOriginalPrice();
```
Comment (pamil) [inline src/Sylius/Component/Core/Calculator/ProductVariantOriginalPriceCalculator.php]: Another solution in order to avoid making the return typehint nullable could be to return the regular price if there's no original price.
